include/fiction/technology/charge_distribution_surface.hpp

```diff
@@ -40,75 +45,720 @@ class charge_distribution_surface<Lyt, false> : public Lyt
+    {
+        auto index = cell_to_index(cell);
+        if (index != -1)
+        {
+            strg->cell_charge[static_cast<uint64_t>(index)] = cs;
+        }
+        else
+        {
+            std::cout << "not part of the charge distribution surface" << std::endl;
+        }
+    }
+
+    /**
+     * This function assigns the given charge state to the cell (accessed by the input index) of the layout.
+     *
+     * @param index The index of the cell to which a state of charge is to be assigned.
+     * @param cs The charge state to be assigned to the cell.
+     */
+    void assign_charge_state_index(const uint64_t& index, const sidb_charge_state& cs) const
+    {
+        strg->cell_charge[index] = cs;
+    }
+
+    /**
+     * Sets the charge state of all SiDBs in the layout to a given charge state.
+     *
+     * @param cs The charge state to be assigned to all the SiDBs.
+     */
+    void set_charge_states(const sidb_charge_state& cs) const
+    {
+        for (uint64_t i = 0u; i < strg->cell_charge.size(); i++)
+        {
+            strg->cell_charge[i] = cs;
+        }
+    }
+
+    /**
+     * Returns the charge state of a cell of the layout at a given index.
+     *
+     * @param index The index of the cell.
+     * @return The charge state of the cell at the given index.
+     */
+    [[nodiscard]] sidb_charge_state get_charge_state_index(const uint64_t& index) const noexcept
+    {
+        if (index < (strg->cell_charge.size()))
+        {
+            return strg->cell_charge[index];
+        }
+        return sidb_charge_state::NONE;
+    }
+
+    /**
+     * Returns the charge state of a given cell.
+     *
+     * @param c cell.
+     * @return The charge state of the given cell.
+     */
+    [[nodiscard]] sidb_charge_state get_charge_state_cell(const typename Lyt::cell& c) const noexcept
+    {
+        if (auto index = cell_to_index(c); index != -1)
+        {
+            return strg->cell_charge[static_cast<uint64_t>(index)];
+        }
+        return sidb_charge_state::NONE;
+    }
+
+    /**
+     * Initializes the distance matrix between all the cells of the layout.
+     */
+    void initialize_distance_matrix() const
+    {
+        strg->dist_mat = std::vector<std::vector<double>>(this->num_cells(), std::vector<double>(this->num_cells(), 0));
+
+        for (uint64_t i = 0u; i < strg->sidb_order.size(); i++)
+        {
+            for (uint64_t j = 0u; j < strg->sidb_order.size(); j++)
+            {
+                strg->dist_mat[i][j] = distance_sidb_pair(strg->sidb_order[i], strg->sidb_order[j]);
+            }
+        }
+    };
+
+    /**
+     * Computes the position of a cell in nanometers.
+     *
+     * @param c The cell to compute the position for.
+     * @return A pair of double values representing the x,y position of the cell in nanometers.
+     */
+    std::pair<double, double> nm_position(const typename Lyt::cell& c) const
```

Comment:
Is this a function that should be in this class? Should it not be a generic technology function that receives the lattice size as params and also distinguishes between coordinate types via `if constexpr`?